service/inspectImages.go

```diff
@@ -0,0 +1,271 @@
+	shell = os.Getenv("SHELL")
+	if shell == "" {
+		i.logger.Info("could not determine shell $SHELL using default `/bin/sh`")
+		shell = "/bin/sh"
+	} else {
+		i.logger.Info("using $SHELL: %s", shell)
+	}
+
+	if !i.io.ExistFile(i.target) {
+		i.logger.Fatal("target %s does not exist", i.target)
+		return errors.New("target " + i.target + " does not exist")
+	}
+
+	var err error
+	if i.io.IsDirectory(i.target) {
+		err = i.processDirectory(shell)
+	} else {
+		err = i.processTgz(i.target, shell)
+	}
+	if err != nil {
+		i.logger.Error("procesing target %s", i.target)
+		return errors.New("procesing target " + i.target)
+	}
+
+	err = i.formatter.Output(i.buffer)
+	if err != nil {
+		i.logger.Error("writing output: %s", err)
+		return err
+	}
+	return nil
+}
+
+func (i *ImagesService) processDirectory(shell string) error {
+	err := os.Chdir(i.target)
+	if err != nil {
+		i.logger.Error("moving to target `%s` folder: %s", i.target, err)
+		return err
+	}
+
+	cmd := exec.Command(shell, "-c", "ls *.tgz")
+	out, err := cmd.Output()
+	if err != nil {
+		i.logger.Error("reading tgz files: %s", err)
+		return err
+	}
+	targets := strings.Split(string(out), "\n")
+	for _, t := range targets {
+		if t == "" {
+			continue
+		}
+		err = i.processTgz(t, shell)
+		if err != nil {
+			return err
+		}
+		//Return to target dir to read next tgz file
+		err = os.Chdir(i.target)
+		if err != nil {
+			i.logger.Error("moving to target `%s` folder: %s", i.target, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (i *ImagesService) processTgz(target string, shell string) error {
+	i.logger.Info("processing target: %s", target)
+	folder, err := i.extractFiles(target)
+	if err != nil {
+		return err
+	}
+
+	appFolder := folder + string(filepath.Separator) + "templates"
+	err = os.Chdir(appFolder)
+	if err != nil {
+		i.logger.Error("moving to templates folder: %s", err)
+		return err
+	}
+
+	output, err := extractImagesTemplates(shell, i.logger)
+	if err != nil {
+		return err
+	}
+
+	err = os.Chdir("../")
+	if err != nil {
+		i.logger.Error("moving %s: %s", folder, err)
+		return err
+	}
+
+	values, err := readValues(i.io, i.logger)
+	if err != nil {
+		return err
+	}
+
+	for _, im := range output {
+		im, err := applyTemplateToImage(im, values, i.logger)
+		if err != nil {
+			return err
+		}
+		t, image := validateAndExtractImage(im)
+		if !t {
+			return errors.New("image not valid")
+		}
+		appendImage(image, &i.buffer)
+	}
+	os.RemoveAll(folder)
+	return nil
+}
+
+func validateAndExtractImage(s string) (bool, image) {
+	var rex = regexp.MustCompile(`^(((?P<o>[A-Za-z0-9_\-.:]+)\/){1,2})?((?P<i>[A-Za-z0-9_\-.]+)\:){1}(?P<v>[A-Za-z0-9_\-.]+){1}`)
+	var im image
+	if rex.MatchString(s) {
+		regexResult := rex.FindAllStringSubmatch(s, -1)[0]
+		h := strings.Split(regexResult[1], "/")
+		hostname := ""
+		if len(h) > 2 {
+			hostname = h[0]
+		}
+		im = image{Hostname: hostname, Organization: regexResult[3], Image: regexResult[5], Version: regexResult[6]}
```

Comment:
Parsing a docker image name is really tricky. As a team we can tell you some horror stories about that...

I would not bother to do that. The purpose of the tool is to dump the full image name (so that I can do `docker pull` or similar against the output).

I would simply drop all this code and KISS.

Reply:
Because images can have varying names (a path or full url) and there can be cases where people put the image, org, and version on different lines or variations of the same line it might be worth using `helm template` to create the kubernetes objects and then use their known data structures to extract the images.

Just a thought.